odoo/addons/base/models/res_device.py

```diff
@@ -0,0 +1,184 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from contextlib import nullcontext
+from datetime import datetime
+import logging
+
+from odoo import api, fields, models, tools
+from odoo.http import GeoIP, request, root
+from odoo.tools import SQL
+from odoo.tools.translate import _
+from .res_users import check_identity
+
+_logger = logging.getLogger(__name__)
+
+
+class ResDeviceLog(models.Model):
+    _name = 'res.device.log'
+    _description = 'Device Log'
+    _rec_name = 'display_name'
+
+    session_identifier = fields.Char("Session Identifier", required=True, index='btree')
+    platform = fields.Char("Platform")
+    browser = fields.Char("Browser")
+    ip_address = fields.Char("IP Address")
+    country = fields.Char("Country")
+    city = fields.Char("City")
+    device_type = fields.Selection([('computer', 'Computer'), ('mobile', 'Mobile')], "Device Type")
+    user_id = fields.Many2one("res.users", index='btree')
+    first_activity = fields.Datetime("First Activity")
+    last_activity = fields.Datetime("Last Activity", index='btree')
+    revoked = fields.Boolean("Revoked", index='btree',
+                            help="""If True, the session file corresponding to this device
+                                    no longer exists on the filesystem.""")
+    is_current = fields.Boolean("Current Device", compute="_compute_is_current")
+    linked_ip_addresses = fields.Text("Linked IP address", compute="_compute_linked_ip_addresses")
+
+    def _compute_display_name(self):
+        for device in self:
+            platform = device.platform or _("Unknown")
+            browser = device.browser or _("Unknown")
+            device.display_name = f"{platform.capitalize()} {browser.capitalize()}"
+
+    def _compute_is_current(self):
+        for device in self:
+            device.is_current = request and request.session.sid.startswith(device.session_identifier)
+
+    def _compute_linked_ip_addresses(self):
+        device_group = self._read_group(
+            domain=[],
```

Comment:
Missing domain, it won't scale. `[('session_identifier', 'in', self.mapped('session_identifier')]` maybe ?

Reply:
Yes, we only need records that respect this domain.